### PR TITLE
Adding rest request to key management service

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
Passing RestRequest to key management service fo provide extra request information to key management service getKey method.
Passing it through GetBlobOptions and PutBlobOpitions.

### CODE REVIEW

File: ambry-api/src/main/java/com/github/ambry/router/GetBlobOptionsBuilder.java [hunk 1]
```diff
@@ -69,6 +71,15 @@ public GetBlobOptionsBuilder resolveRangeOnEmptyBlob(boolean resolveRangeOnEmpty
     return this;
   }
 
+  /**
+   * @param restRequest The {@link RestRequest} that triggered this get operationt.
```
Comment (cgtz) [inline ambry-api/src/main/java/com/github/ambry/router/GetBlobOptionsBuilder.java:75]: operationt -> operation

File: ambry-api/src/main/java/com/github/ambry/router/GetBlobOptions.java [hunk 2]
```diff
@@ -115,6 +119,10 @@ public boolean hasBlobSegmentIdx() {
     return blobSegmentIdx != NO_BLOB_SEGMENT_IDX_SPECIFIED;
   }
 
+  public RestRequest getRestRequest() {
```
Comment (cgtz) [inline ambry-api/src/main/java/com/github/ambry/router/GetBlobOptions.java:122]: For consistency, let's just add a javadoc to this method, even if its somewhat redundant with the builder.

File: ambry-api/src/main/java/com/github/ambry/router/KeyManagementService.java [hunk 3]
```diff
@@ -47,21 +48,23 @@
...
    * Fetches the key associated with the pair of AccountId and ContainerId. User is expected to have
    * registered using {@link #register(short, short)} for this pair before fetching keys.
+   * @param restRequest the {@link RestRequest} to use
    * @param accountId refers to the id of the {@link Account} for which key is expected
    * @param containerId refers to the id of the {@link Container} for which key is expected
    * @return T the key associated with the accountId and containerId
    * @throws {@link GeneralSecurityException} on KMS unavailability or if key is not registered
    */
-  T getKey(short accountId, short containerId) throws GeneralSecurityException;
+  T getKey(RestRequest restRequest, short accountId, short containerId) throws GeneralSecurityException;
```
Comment (cgtz) [inline ambry-api/src/main/java/com/github/ambry/router/KeyManagementService.java:57]: Let's document that a KMS impl should be prepared to handle the case where `restRequest` is null.

File: ambry-router/src/test/java/com/github/ambry/router/MockKeyManagementService.java [hunk 4]
```diff
@@ -30,11 +31,12 @@
   }
 
   @Override
-  public SecretKeySpec getKey(short accountId, short containerId) throws GeneralSecurityException {
+  public SecretKeySpec getKey(RestRequest restRequest, short accountId, short containerId)
```
Comment (cgtz) [inline ambry-router/src/test/java/com/github/ambry/router/MockKeyManagementService.java:34]: Could we make one of the router tests pass in a non-null restRequest in the options and validate that it gets passed correctly to the KMS impl

**Review by cgtz — COMMENTED** [review]
LGTM, mainly one comment about testing

**Review by zzmao — APPROVED** [review]
LGTM

**Pull request merged**